File: toolkit-core/core-qbean-api/src/main/java/net/sunshow/toolkit/core/qbean/api/search/PageData.java
package net.sunshow.toolkit.core.qbean.api.search;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public class PageData<E> {

    /* 回写请求时的分页设置 */
    private int page;
    private int pageSize;

    /* 满足条件的记录总数 */
    private long total = 0;

    private Collection<E> data;

    public PageData(int page, int pageSize, Collection<E> data, long total) {
        this.page = page;
        this.pageSize = pageSize;
        if (data == null) {
            this.data = new ArrayList<>();
        } else {
            this.data = data;
        }
        this.total = total;
    }

    public PageData(int page, int pageSize, Collection<E> data) {
        this(page, pageSize, data, 0);
    }

    public PageData(int page, int pageSize) {
        this(page, pageSize, null);
    }

    public PageData() {
    }

    public <T> PageData<T> map(Function<E, T> mapper) {
        List<T> mapList = this.getData().stream()
                .map(mapper)
                .collect(Collectors.toList());
        return new PageData<>(this.getPage(), this.getPageSize(), mapList, this.getTotal());
    }

    public int getCount() {
        if (data == null) {
            return 0;
        }

        return data.size();
    }

    public int getPageTotal() {
        if (total <= 0) {
            return 0;
        }

        return (int) Math.ceil((double) total / pageSize);
    }

    public int getPage() {
        return page;
    }

    public void setPage(int page) {
        this.page = page;
    }

    public int getPageSize() {
        return pageSize;
    }

    public void setPageSize(int pageSize) {
        this.pageSize = pageSize;
    }

    public long getTotal() {
        return total;
    }

    public void setTotal(long total) {
        this.total = total;
    }

    public Collection<E> getData() {
        if (data == null) {
            data = new ArrayList<>();
        }
        return data;
    }

    public void setData(Collection<E> data) {
        this.data = data;
    }
}

File: toolkit-core/core-qbean-api/src/main/kotlin/net/sunshow/toolkit/core/qbean/api/search/PageData.kt
package net.sunshow.toolkit.core.qbean.api.search

import java.util.function.Function
import kotlin.math.ceil

open class PageData<E> {
    /* 回写请求时的分页设置 */
    var page: Int = 0
        private set
    var pageSize: Int = 0
        private set

    /* 满足条件的记录总数 */
    var total: Long = 0
        private set

    var data: Collection<E> = emptyList()

    constructor(page: Int, pageSize: Int, data: Collection<E>, total: Long) {
        this.page = page
        this.pageSize = pageSize
        this.data = data
        this.total = total
    }

    constructor(page: Int, pageSize: Int, data: Collection<E>) : this(
        page,
        pageSize,
        data,
        data.size.toLong()
    )

    constructor(page: Int, pageSize: Int) : this(page, pageSize, emptyList())

    constructor()

    open fun <T> map(mapper: Function<E, T>): PageData<T> {
        val mapList = this.data
            .map {
                mapper.apply(it)
            }
        return PageData(this.page, this.pageSize, mapList, this.total)
    }

    open fun getCount(): Int {
        return data.size
    }

    open fun getPageTotal(): Int {
        if (total <= 0) {
            return 0
        }

        return ceil(total.toDouble() / pageSize).toInt()
    }
}
